### PR TITLE
k8s: Add support for 1.31.0

### PR DESCRIPTION
**Pull request by christarazi**
- **go.mod: Bump controller-tools to v0.16.1**
- **docs: Update kindest/node instructions for K8s upgrade**
- **k8s: Add support for 1.31.0**

### CODE REVIEW

**Review by learnitall — APPROVED** [review]
Docs changes look good, thanks!

**Review by julianwiedmann — APPROVED** [review]

**Review by viktor-kurchenko — APPROVED** [review]

**Review by marseel — APPROVED** [review]

**Review by tklauser — APPROVED** [review]
The ["Build commits" workflow failure](https://github.com/cilium/cilium/actions/runs/10466497702/job/28983469640?pr=34463) looks legitimate. Probably the commits need to be reordered/squashed?

Otherwise LGTM, thanks!

**Review by squeed — APPROVED** [review]

**Comment by christarazi** [comment]
CI should be fixed now. There was a breaking change in the K8s code generation dependencies that was tricky to nail down. See first commit msg for more details.

**Comment by christarazi** [comment]
Still seeing some CI failures, investigating. For now I'll put it into draft.

**Comment by sayboras** [comment]
Ingress and Gateway API failures will be fixed after the below PRs are merged

https://github.com/cilium/cilium/pull/34575
https://github.com/cilium/cilium/pull/34576

File: .github/actions/ginkgo/main-k8s-versions.yaml [hunk 1]
```diff
@@ -1,6 +1,13 @@
 # This file contains which kernel versions should be run with which k8s versions
 ---
 include:
+  - k8s-version: "1.31"
+    ip-family: "dual"
+    # renovate: datasource=docker
+    kube-image: "quay.io/cilium/kindest-node:v1.31.0-rc.1@sha256:892dad427fa73e80ff16e26e287fae4d45aeaf8055206936e42f82b1dfedf2a4"
```
Comment (aanm) [inline .github/actions/ginkgo/main-k8s-versions.yaml]: `quay.io/cilium/kindest-node:v1.31.0@sha256:7e3cb518db362bee6a507dd5c9aa288a979ee8899be88231cdd4e9ca4079ac30`

**Review by aanm — CHANGES_REQUESTED** [review]

**Comment by sayboras** [comment]
> Ingress and Gateway API failures will be fixed after the below PRs are merged
> 
> #34575 #34576

I have rebased with main branch, also fix the conflict with main (due to my changes) and force pushed.

**Comment by sayboras** [comment]
Force pushed to fix the related CI failures + rebase with main branch. 

All CI jobs seem green now, marking this ready for review if you don't mind.

**Review by sayboras — APPROVED** [review]

**Review by aanm — APPROVED** [review]

**Comment by julianwiedmann** [comment]
Given that this touches CI - should it have been tested from a cilium/cilium branch, instead of a forked repo?

**Comment by mhofstetter** [comment]
> Given that this touches CI - should it have been tested from a cilium/cilium branch, instead of a forked repo?

Probably yes. It seems as `/ci-ginkgo` fails for k8s v1.27 after removing the mapping k8s v1.27 to kind image

```
Error: cannot pull image quay.io/lvh-images/kind:: invalid reference format
```

Trying to fix it over here: https://github.com/cilium/cilium/pull/34773

**Comment by christarazi** [comment]
> Given that this touches CI - should it have been tested from a cilium/cilium branch, instead of a forked repo?

Hm, that's something I didn't think about. I'll update the docs to explicitly mention this detail so we can avoid this situation on the next K8s bump.

**Pull request merged**